src/realm/sync/noinst/client_impl_base.hpp

```diff
@@ -35,7 +35,32 @@ namespace sync {
 //
 // `protocol` is included for convenience, even though it is not strictly part
 // of an endpoint.
-using ServerEndpoint = std::tuple<ProtocolEnvelope, std::string, network::Endpoint::port_type>;
+
+struct ServerEndpoint {
+    ProtocolEnvelope envelope;
+    std::string address;
+    network::Endpoint::port_type port;
+    std::string user_id;
+    SyncServerMode server_mode = SyncServerMode::PBS;
+
+private:
+    auto to_tuple() const
+    {
+        return std::make_tuple(server_mode, envelope, std::ref(address), port, std::ref(user_id));
+    }
+
+public:
+    friend inline bool operator==(const ServerEndpoint& lhs, const ServerEndpoint& rhs)
+    {
+        return lhs.to_tuple() == rhs.to_tuple();
+    }
+
+
+    friend inline bool operator<(const ServerEndpoint& lhs, const ServerEndpoint& rhs)
```

Comment:
do we use `ServerEndpoint`s as keys in a map or why do we need this? I find it a bit strange..

Reply:
Yes - the map is defined here: https://github.com/realm/realm-core/blob/master/src/realm/sync/noinst/client_impl_base.hpp#L209